Major revision and update of the syllabus (#63)
merely-useful/r-rse: chapters/01-syllabus.Rmd

# Course Syllabus {#syllabus}

Computational skills are essential in the modern research environment, which
increasingly have massive datasets or use high performance computing servers,
along with a trend for a greater demand that scientific studies be
reproducible and open. This book will teach you the tools, skills, and knowledge
necessary to make research software that is open, reproducible, sustainable, and
more rigorous. (TODO: other word?) First and foremost, the main learning goal of
this course is that you learn to **create and develop an R package**.
More specifically, the learning objectives are to:

- Explain what an R package is, how it is structured, and how to set one up.

> - List and describe what some best (or good enough) practices are in package

Reviewer: Why only some best practices? If I were a reader, I'd be wondering what I'm missing.

development and how to implement them. 
- Explain what a code style guide is, why it is important, and how to write code
that adheres to the style guide.
- Describe why user-friendly and well-written documentation is crucial to

> building a usable package and some steps to take to achieve this type of

Reviewer: ```suggestion
building a usable package and some steps to achieve this type of
```

documentation.
- Cover what CRAN is, what policies and quality-control steps it has for packages,
and how to prepare a package for submission to CRAN so that it passes these checks
and requirements.

> Aside from the direct technical aspects of building an R package, the course

Reviewer: ```suggestion
Aside from the technical aspects of building an R package, the course
```

will also cover topics that enhance and improve the process of package development.
These learning objectives are to:

> - Describe and apply effective workflows and project management tools that help
> to coordinate the development of software, particularly for software that is

Reviewer: ```suggestion
- Describe and apply effective workflows and project management tools that
coordinate the development of software, particularly for software that is
```

complex or when working in a team.
- Outline (and optionally practice) how to work productively in a small team,
that is inclusive and welcoming.

> - Explain the importance of and learning to design software and documentation
> that involves the community, such as by gaining potential contributors.

Reviewer: ```suggestion
- Explain the importance of and learning to design software and documentation
involving community contributors.
```

> TODO: Include objectives about having ORCID and getting DOIs?

Reviewer: I'm not personally in favor of creating an objective for ORCID and DOIs, as that will require us to go through the process ourselves. I think it *should* be included in the narrative, but as a discussion point as another option.

Overall, we'll be addressing the questions:

- What is an R package and why should I create one?
- How can I build an R package that is re-usable, well-built, and well-documented?
- What is CRAN and should I submit my package to it? If so, how do I do it?
- How can I work effectively as a team on building an R package or to

> contribute to an existing package?

Reviewer: I'm also not sure I like listing this set of summary objectives. I think they should be converted from questions to statements, and added to the section below ("by the end of the course")

By the end of the course, learners are expected to have the skills and knowledge
necessary to build their own packages and to, if relevant, submit these packages
to CRAN.

> TODO: Include bioconductor?

Reviewer: I'm also not in favor of including bioconductor (except as a brief mention in the chapter); it involves adding too much extra detailed conceptual framework about how installation works, submissions to the repository, and why the objects are so big and weird.

Reviewer 2: Yeah, will we do something specifically bio-related that requires knowing bioconductor?

both to check them and to build on them. It's important to note that science can
be open but not reproducible and reproducible but not open.
- [Sustainable software][sustainable-software] is when software is developed in
a way that makes it easier for other people to extend and/or maintain it, rather
than to replace it. Sustainability isn't just about the software itself, it also
depends on the skills and culture of its users. For instance, if a software
package is being maintained by a couple of post-docs who are being paid a
fraction of what they could earn in industry and have no realistic hope of
promotion because their field doesn't value tool building, it doesn't matter
whether the package is well tested and easy to install: sooner or later
(probably sooner) it will become [abandonware][abandonware].

## Overview of the packages created throughout course

Throughout the book, you will build several packages. There is the main
one (`zipfs`) and 2 or 3 project assignment packages (see the assignment
descriptions in Appendix \@ref(course-assignments)). For two of the packages,
we have already built them so you can use them as an example and a reference

> (`zipfs` and `kenyaweather`).

Author: Hmm, not sure this is the best place for it, but also not sure right now where a better place would be. Adding for now, we can decide later.

- The main package used throughout the book is the Zipf's Law package `zipfs`.
This package is used for the code-alongs, the actual content, and most of the
exercises.
- The example package for the first project assignment is the Weather in Kenya
package `kenyaweather`. This package is used as a reference when you work on
your package for the first project assignment.
- A currently existing package that you'll contribute to, which is the second
project assignment.

> - (Optional) A group-based package, which is the third project assignment.

Reviewer 3: ```suggestion
- Optionally, you will create a package in a group for the third project assignment.
```
